rdtools/normalization.py

```diff
@@ -209,175 +258,524 @@ def normalize_with_sapm(energy, sapm_kws):
+                         'DatetimeIndex')
+
+    if len(power) <= 1:
+        # just one value, doesn't make sense to interpolate or trapz aggregate.
+        # use the index frequency to determine the appropriate timescale
+        if target_frequency is None:
+            if power.index.freq is None:
+                raise ValueError('Could not determine period of input power')
+
+            target_frequency = power.index.freq
+        # just raise if it's a non-fixed frequency
+        interval_length_ns = \
+            pd.tseries.frequencies.to_offset(target_frequency).nanos
+
+        energy = power * interval_length_ns / 1e9 / 3600  # ns to s to h
+        energy.name = 'energy_Wh'
+        return energy
+
+    t_steps = t_step_nanoseconds(power)
+    median_step_ns = t_steps.median()
+
+    if target_frequency is None:
+        # 'N' is the Pandas offset alias for ns
+        target_frequency = str(int(median_step_ns)) + 'N'
+
+    if max_timedelta is None:
+        max_interval_nanoseconds = median_step_ns
+    else:
+        max_interval_nanoseconds = max_timedelta.total_seconds() * 10.0**9
+
+    try:
+        freq_interval_size_ns = \
+            pd.tseries.frequencies.to_offset(target_frequency).nanos
+    except ValueError as e:
+        if 'is a non-fixed frequency' in str(e):
+            temp_ind = pd.date_range(power.index[0],
+                                     power.index[-1],
+                                     freq=target_frequency)
+            temp_series = pd.Series(data=1, index=temp_ind)
+            temp_diffs = t_step_nanoseconds(temp_series)
+            freq_interval_size_ns = temp_diffs.median()
+        else:
+            raise
+
+    # Upsampling case
+    if freq_interval_size_ns <= median_step_ns:
+        resampled = interpolate(power, target_frequency, max_timedelta)
+
+        moving_average = (resampled + resampled.shift()) / 2.0
+
+        energy = moving_average * t_step_nanoseconds(moving_average) \
+                    / 10.0**9 / 3600.0
+
+        # Drop first row with work around for pandas issue #18031
+        if energy.index.tz is None:
+            energy = energy.drop(energy.index[0])
+        else:
+            tz = str(energy.index.tz)
+            energy.index = energy.index.tz_convert('UTC')
+            energy = energy.drop(energy.index[0])
+            energy.index = energy.index.tz_convert(tz)
+
+    # Downsampling case
+    elif freq_interval_size_ns > median_step_ns:
+        energy = trapz_aggregate(power, target_frequency, max_timedelta)
+
+    # Set the frequency if we can
+    try:
+        energy.index.freq = pd.infer_freq(energy.index)
+    except ValueError:
+        pass
+
+    # enforce max_timedelta
+    t_steps = t_steps.reindex(energy.index, method='backfill')
+    energy.loc[t_steps > max_interval_nanoseconds] = np.nan
+
+    energy.name = 'energy_Wh'
+
+    return energy
+
+
+def trapz_aggregate(time_series, target_frequency, max_timedelta=None):
```

Comment:
seems like an opportunity to use `time_series.rolling(window).apply(np.trapz)` if `window = some_function(target_frequence)` can be done.

Reply:
Thank you for the suggestion. Ultimately, this function got more complicated to handle either a a trapezoidal or rectangular sum in #217